crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -0,0 +1,1714 @@
+                    .map_err(InstantiationError::Resource)?,
+            );
+        }
+
+        let mut dropped_data = instance.dropped_data.borrow_mut();
+        dropped_data.clear();
+        dropped_data.resize(module.passive_data.len());
+
+        Ok(())
+    }
+
+    fn set_instance_tables(
+        instance: &mut Instance,
+        mut tables: BasePointerIterator,
+        max_elements: u32,
+    ) -> Result<(), InstantiationError> {
+        let module = instance.module.as_ref();
+
+        instance.tables.clear();
+
+        for plan in (&module.table_plans.values().as_slice()[module.num_imported_tables..]).iter() {
+            let base = tables.next().unwrap();
+
+            // Make the table data accessible
+            if unsafe { !make_accessible(base, max_elements as usize * max_table_element_size()) } {
+                return Err(InstantiationError::Resource(
+                    "failed to make instance memory accessible".into(),
+                ));
+            }
+
+            instance
+                .tables
+                .push(Table::new_static(plan, base, max_elements));
+        }
+
+        let mut dropped_elements = instance.dropped_elements.borrow_mut();
+        dropped_elements.clear();
+        dropped_elements.resize(module.passive_elements.len());
+
+        Ok(())
+    }
+}
+
+impl Drop for InstancePool {
+    fn drop(&mut self) {
+        unsafe {
+            for i in 0..self.max_instances {
+                let ptr = self.mapping.as_mut_ptr().add(i * self.instance_size) as *mut Instance;
+                std::ptr::drop_in_place(ptr);
+            }
+        }
+    }
+}
+
+/// Represents a pool of WebAssembly linear memories.
+///
+/// A linear memory is divided into accessible pages and guard pages.
+///
+/// Each instance index into the pool returns an iterator over the base addresses
+/// of the instance's linear memories.
+///
+///
+/// The userfault handler relies on how memories are stored in the mapping,
+/// so make sure the uffd implementation is kept up-to-date.
+#[derive(Debug)]
+struct MemoryPool {
+    mapping: Mmap,
+    memory_size: usize,
+    max_memories: usize,
+    max_instances: usize,
+    max_wasm_pages: u32,
+}
+
+impl MemoryPool {
+    fn new(module_limits: &ModuleLimits, instance_limits: &InstanceLimits) -> Result<Self, String> {
+        let memory_size = usize::try_from(instance_limits.memory_reservation_size)
+            .map_err(|_| "memory reservation size exceeds addressable memory".to_string())?;
+
+        debug_assert!(
+            memory_size % region::page::size() == 0,
+            "memory size {} is not a multiple of system page size",
+            memory_size
+        );
+
+        let max_instances = instance_limits.count as usize;
+        let max_memories = module_limits.memories as usize;
+
+        let allocation_size = memory_size
+            .checked_mul(max_memories)
+            .and_then(|c| c.checked_mul(max_instances))
```

Comment:
The reason the memory and table pools don't maintain their own free lists is that the uffd implementation can very quickly go from faulting address to the owning instance based on some simple arithmetic without any extra bookkeeping; but I don't think that's a strong enough reason to not be more flexible in this implementation.

We'd need a table that maps between memory/table pool index to instance that it was allocated for, which isn't that big a deal to maintain.

I'll see about addressing this.

Reply:
I'm waffling on this a bit over time myself. I'm not actually sure that this is a good idea, because if you want to *guarantee* that each instance can have a memory (or two memories) then you have to do what your implementation does today anyway (basically pre-assign memories per instance). Having a few extra memories so some modules can have 2 memories and most others can have one only works when everything is cooperative.

I feel like the cooperative case doesn't really happen that much if you're using the pooling allocator?

I think this should probably actually stay as-is where the limits are guaranteed to be available for each instance, and we can always revisit this in the future if necessary.

Reply:
Resolving as we're going to keep the current design where we guarantee an instance can allocate provided 1) the module conforms to the module limits and 2) the maximum number of instances hasn't been reached.

We can address this in the future if this needs to change.